cpp/src/Ice/HashUtil.h

```diff
@@ -7,50 +7,60 @@
 
 #include <cstddef>
 #include <cstdint>
+#include <functional>
 #include <map>
 #include <string>
 #include <vector>
 
+#include "EndpointI.h"
+
+// All these functions are for hashing proxies and endpoints. They are not general purpose.
+
 namespace IceInternal
 {
-    inline void hashAdd(std::int32_t& hashCode, std::int32_t value)
+    inline void hashAdd(std::size_t& hashCode, std::size_t value)
     {
-        hashCode = ((hashCode << 5) + hashCode) ^ static_cast<std::int32_t>(2654435761u) * value;
+        hashCode = ((hashCode << 5) + hashCode) ^ 2654435761u * value;
     }
 
-    inline void hashAdd(std::int32_t& hashCode, bool value)
+    inline void hashAdd(std::size_t& hashCode, std::int32_t value)
+    {
+        hashAdd(hashCode, static_cast<std::size_t>(value));
+    }
+
+    inline void hashAdd(std::size_t& hashCode, bool value)
     {
         hashCode = ((hashCode << 5) + hashCode) ^ (value ? 1 : 0);
     }
 
-    inline void hashAdd(std::int32_t& hashCode, std::byte value)
+    template<class T> inline void hashAdd(std::size_t& hashCode, const T& value)
     {
-        // TODO: better hash function for std::byte
-        hashAdd(hashCode, static_cast<std::int32_t>(value));
+        hashAdd(hashCode, std::hash<T>{}(value));
     }
 
-    inline void hashAdd(std::int32_t& hashCode, const std::string& value)
+    inline void hashAdd(std::size_t& hashCode, const std::vector<EndpointIPtr>& seq)
     {
```

Comment:
We could add the container's size to the hash, when hashing a collection.

Reply:
I'd rather keep the same logic as in C#.

In C#, since we treat null collection = empty collection, and we skip nulls, we can't add the size of the collection to the hash.